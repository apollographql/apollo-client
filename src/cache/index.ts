export { Transaction, ApolloCache } from './core/cache';
export { Cache } from './core/types/Cache';
export { DataProxy } from './core/types/DataProxy';
export { MissingFieldError } from './core/types/common';

export {
  Reference,
  isReference,
  makeReference,
} from '../utilities/graphql/storeUtils';

export {
  InMemoryCache,
  InMemoryCacheConfig,
  ReactiveVar,
} from './inmemory/inMemoryCache';

export {
  defaultDataIdFromObject,
  TypePolicies,
  TypePolicy,
  FieldPolicy,
  FieldReadFunction,
  FieldMergeFunction,
  FieldFunctionOptions,
  PossibleTypesMap,
} from './inmemory/policies';

export * from './inmemory/types';
