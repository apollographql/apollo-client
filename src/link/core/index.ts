export { empty } from './empty';
export { from } from './from';
export { split } from './split';
export { concat } from './concat';
export { execute } from './execute';
export { ApolloLink } from './ApolloLink';

export * from './types';
