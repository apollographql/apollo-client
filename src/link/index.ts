export { empty } from "./core/empty.js";
export { from } from "./core/from.js";
export { split } from "./core/split.js";
export { concat } from "./core/concat.js";
export { execute } from "./core/execute.js";
export { ApolloLink } from "./core/ApolloLink.js";

export type {
  ApolloPayloadResult,
  DocumentNode,
  ExecuteContext,
  ExecutionPatchIncrementalResult,
  ExecutionPatchInitialResult,
  ExecutionPatchResult,
  FetchResult,
  GraphQLRequest,
  IncrementalPayload,
  NextLink,
  Operation,
  OperationContext,
  Path,
  RequestHandler,
  SingleExecutionResult,
} from "./core/types.js";
