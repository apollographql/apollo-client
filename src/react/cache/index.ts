export type { SuspenseCacheOptions } from './SuspenseCache.js';
export { getSuspenseCache } from './getSuspenseCache.js';

import { SuspenseCache as RealSuspenseCache } from './SuspenseCache.js';

// TODO: remove export with release 3.8
// replace with
// export type { SuspenseCache } from './SuspenseCache.js';
/**
 * @deprecated 
 * It is no longer necessary to create a `SuspenseCache` instance and pass it into the `ApolloProvider`.
 * Please remove this code from your application.
 *  
 * This export will be removed with the final 3.8 release.
 */
export class SuspenseCache extends RealSuspenseCache {
  constructor(){
    super();
    // throwing an error here instead of using invariant - we do not want this error
    // message to be link-ified, but to directly show up as bold as possible
    throw new Error(
      'It is no longer necessary to create a `SuspenseCache` instance and pass it into the `ApolloProvider`.\n' + 
      'Please remove this code from your application. \n\n' + 
      'This export will be removed with the final 3.8 release.'
    );
  }
}