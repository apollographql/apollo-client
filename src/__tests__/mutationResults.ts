import { cloneDeep } from 'lodash';
import gql from 'graphql-tag';

import { ApolloClient } from '../core';
import { InMemoryCache } from '../cache';
import { ApolloLink } from '../link/core';
import { Observable, ObservableSubscription as Subscription } from '../utilities';
import { itAsync, subscribeAndCount, mockSingleLink } from '../testing';

describe('mutation results', () => {
  const query = gql`
    query todoList {
      todoList(id: 5) {
        id
        todos {
          id
          text
          completed
        }
        filteredTodos: todos(completed: true) {
          id
          text
          completed
        }
      }
      noIdList: todoList(id: 6) {
        id
        todos {
          text
          completed
        }
      }
    }
  `;

  const queryWithTypename = gql`
    query todoList {
      todoList(id: 5) {
        id
        todos {
          id
          text
          completed
          __typename
        }
        filteredTodos: todos(completed: true) {
          id
          text
          completed
          __typename
        }
        __typename
      }
      noIdList: todoList(id: 6) {
        id
        todos {
          text
          completed
          __typename
        }
        __typename
      }
    }
  `;

  const result: any = {
    data: {
      __typename: 'Query',
      todoList: {
        __typename: 'TodoList',
        id: '5',
        todos: [
          {
            __typename: 'Todo',
            id: '3',
            text: 'Hello world',
            completed: false,
          },
          {
            __typename: 'Todo',
            id: '6',
            text: 'Second task',
            completed: false,
          },
          {
            __typename: 'Todo',
            id: '12',
            text: 'Do other stuff',
            completed: false,
          },
        ],
        filteredTodos: [],
      },
      noIdList: {
        __typename: 'TodoList',
        id: '7',
        todos: [
          {
            __typename: 'Todo',
            text: 'Hello world',
            completed: false,
          },
          {
            __typename: 'Todo',
            text: 'Second task',
            completed: false,
          },
          {
            __typename: 'Todo',
            text: 'Do other stuff',
            completed: false,
          },
        ],
      },
    },
  };

  function setupObsQuery(
    reject: (reason: any) => any,
    ...mockedResponses: any[]
  ) {
    const client = new ApolloClient({
      link: mockSingleLink({
        request: { query: queryWithTypename } as any,
        result,
      }, ...mockedResponses).setOnError(reject),
      cache: new InMemoryCache({
        dataIdFromObject: (obj: any) => {
          if (obj.id && obj.__typename) {
            return obj.__typename + obj.id;
          }
          return null;
        },
        // Passing an empty map enables warnings about missing fields:
        possibleTypes: {},
      }),
    });

    return {
      client,
      obsQuery: client.watchQuery({
        query,
        notifyOnNetworkStatusChange: false,
      }),
    };
  }

  function setupDelayObsQuery(
    reject: (reason: any) => any,
    delay: number,
    ...mockedResponses: any[]
  ) {
    const client = new ApolloClient({
      link: mockSingleLink({
        request: { query: queryWithTypename } as any,
        result,
        delay,
      }, ...mockedResponses).setOnError(reject),
      cache: new InMemoryCache({
        dataIdFromObject: (obj: any) => {
          if (obj.id && obj.__typename) {
            return obj.__typename + obj.id;
          }
          return null;
        },
        // Passing an empty map enables warnings about missing fields:
        possibleTypes: {},
      }),
    });

    return {
      client,
      obsQuery: client.watchQuery({
        query,
        notifyOnNetworkStatusChange: false,
      }),
    };
  }

  itAsync('correctly primes cache for tests', (resolve, reject) => {
    const { client, obsQuery } = setupObsQuery(reject);
    return obsQuery.result().then(
      () => client.query({ query })
    ).then(resolve, reject);
  });

  itAsync('correctly integrates field changes by default', (resolve, reject) => {
    const mutation = gql`
      mutation setCompleted {
        setCompleted(todoId: "3") {
          id
          completed
          __typename
        }
        __typename
      }
    `;

    const mutationResult = {
      data: {
        __typename: 'Mutation',
        setCompleted: {
          __typename: 'Todo',
          id: '3',
          completed: true,
        },
      },
    };

    const { client, obsQuery } = setupObsQuery(reject, {
      request: { query: mutation },
      result: mutationResult,
    });

    return obsQuery.result().then(() => {
      return client.mutate({ mutation });
    }).then(() => {
      return client.query({ query });
    }).then((newResult: any) => {
      expect(newResult.data.todoList.todos[0].completed).toBe(true);
    }).then(resolve, reject);
  });

  itAsync('correctly integrates field changes by default with variables', (resolve, reject) => {
    const query = gql`
      query getMini($id: ID!) {
        mini(id: $id) {
          id
          cover(maxWidth: 600, maxHeight: 400)
          __typename
        }
      }
    `;
    const mutation = gql`
      mutation upload($signature: String!) {
        mini: submitMiniCoverS3DirectUpload(signature: $signature) {
          id
          cover(maxWidth: 600, maxHeight: 400)
          __typename
        }
      }
    `;

    const link = mockSingleLink({
      request: {
        query,
        variables: { id: 1 },
      } as any,
      delay: 100,
      result: {
        data: { mini: { id: 1, cover: 'image', __typename: 'Mini' } },
      },
    }, {
      request: {
        query: mutation,
        variables: { signature: '1234' },
      } as any,
      delay: 150,
      result: {
        data: { mini: { id: 1, cover: 'image2', __typename: 'Mini' } },
      },
    }).setOnError(reject);

    interface Data {
      mini: { id: number; cover: string; __typename: string };
    }
    const client = new ApolloClient({
      link,
      cache: new InMemoryCache({
        dataIdFromObject: (obj: any) => {
          if (obj.id && obj.__typename) {
            return obj.__typename + obj.id;
          }
          return null;
        },
      }),
    });

    const obs = client.watchQuery<Data>({
      query,
      variables: { id: 1 },
      notifyOnNetworkStatusChange: false,
    });

    let count = 0;
    obs.subscribe({
      next: result => {
        if (count === 0) {
          client.mutate({ mutation, variables: { signature: '1234' } });
          expect(result.data!.mini.cover).toBe('image');

          setTimeout(() => {
            if (count === 0)
              reject(
                new Error('mutate did not re-call observable with next value'),
              );
          }, 250);
        }
        if (count === 1) {
          expect(result.data!.mini.cover).toBe('image2');
          resolve();
        }
        count++;
      },
      error: reject,
    });
  });

  itAsync("should warn when the result fields don't match the query fields", (resolve, reject) => {
    let handle: any;
    let subscriptionHandle: Subscription;

    const queryTodos = gql`
      query todos {
        todos {
          id
          name
          description
          __typename
        }
      }
    `;

    const queryTodosResult = {
      data: {
        todos: [
          {
            id: '1',
            name: 'Todo 1',
            description: 'Description 1',
            __typename: 'todos',
          },
        ],
      },
    };

    const mutationTodo = gql`
      mutation createTodo {
        createTodo {
          id
          name
          # missing field: description
          __typename
        }
      }
    `;

    const mutationTodoResult = {
      data: {
        createTodo: {
          id: '2',
          name: 'Todo 2',
          __typename: 'createTodo',
        },
      },
    };

    const { client, obsQuery } = setupObsQuery(
      reject,
      {
        request: { query: queryTodos },
        result: queryTodosResult,
      },
      {
        request: { query: mutationTodo },
        result: mutationTodoResult,
      },
    );

    return obsQuery.result().then(() => {
      // we have to actually subscribe to the query to be able to update it
      return new Promise(resolve => {
        handle = client.watchQuery({ query: queryTodos });
        subscriptionHandle = handle.subscribe({
          next(res: any) {
            resolve(res);
          },
        });
      });
    }).then(() => client.mutate({
      mutation: mutationTodo,
      updateQueries: {
        todos: (prev, { mutationResult }) => {
          const newTodo = (mutationResult as any).data.createTodo;
          const newResults = {
            todos: [...(prev as any).todos, newTodo],
          };
          return newResults;
        },
      },
    })).then(
      () => {
        subscriptionHandle.unsubscribe();
        fail("should have errored");
      },
      error => {
        subscriptionHandle.unsubscribe();
        expect(error.message).toMatch(/Missing field 'description' /);
      },
    ).then(resolve, reject);
  });

  describe('updateQueries', () => {
    const mutation = gql`
      mutation createTodo {
        # skipping arguments in the test since they don't matter
        createTodo {
          id
          text
          completed
          __typename
        }
        __typename
      }
    `;

    const mutationResult = {
      data: {
        __typename: 'Mutation',
        createTodo: {
          id: '99',
          __typename: 'Todo',
          text: 'This one was created with a mutation.',
          completed: true,
        },
      },
    };

    itAsync('analogous of ARRAY_INSERT', (resolve, reject) => {
      let subscriptionHandle: Subscription;
      const { client, obsQuery } = setupObsQuery(reject, {
        request: { query: mutation },
        result: mutationResult,
      });

      return obsQuery.result().then(() => {
        // we have to actually subscribe to the query to be able to update it
        return new Promise(resolve => {
          const handle = client.watchQuery({ query });
          subscriptionHandle = handle.subscribe({
            next(res) {
              resolve(res);
            },
          });
        });
      }).then(() => client.mutate({
        mutation,
        updateQueries: {
          todoList: (prev, options) => {
            const mResult = options.mutationResult as any;
            expect(mResult.data.createTodo.id).toBe('99');
            expect(mResult.data.createTodo.text).toBe(
              'This one was created with a mutation.',
            );
            const state = cloneDeep(prev) as any;
            state.todoList.todos.unshift(mResult.data.createTodo);
            return state;
          },
        },
      })).then(
        () => client.query({ query })
      ).then((newResult: any) => {
        subscriptionHandle.unsubscribe();

        // There should be one more todo item than before
        expect(newResult.data.todoList.todos.length).toBe(4);

        // Since we used `prepend` it should be at the front
        expect(newResult.data.todoList.todos[0].text).toBe(
          'This one was created with a mutation.',
        );
      }).then(resolve, reject);
    });

    itAsync('does not fail if optional query variables are not supplied', (resolve, reject) => {
      let subscriptionHandle: Subscription;
      const mutationWithVars = gql`
        mutation createTodo($requiredVar: String!, $optionalVar: String) {
          createTodo(requiredVar: $requiredVar, optionalVar: $optionalVar) {
            id
            text
            completed
            __typename
          }
          __typename
        }
      `;

      // the test will pass if optionalVar is uncommented
      const variables = {
        requiredVar: 'x',
        // optionalVar: 'y',
      };
      const { client, obsQuery } = setupObsQuery(reject, {
        request: {
          query: mutationWithVars,
          variables,
        },
        result: mutationResult,
      });

      return obsQuery.result().then(() => {
        // we have to actually subscribe to the query to be able to update it
        return new Promise(resolve => {
          const handle = client.watchQuery({
            query,
            variables,
          });
          subscriptionHandle = handle.subscribe({
            next(res) {
              resolve(res);
            },
          });
        });
      }).then(() => client.mutate({
        mutation: mutationWithVars,
        variables,
        updateQueries: {
          todoList: (prev, options) => {
            const mResult = options.mutationResult as any;
            expect(mResult.data.createTodo.id).toBe('99');
            expect(mResult.data.createTodo.text).toBe(
              'This one was created with a mutation.',
            );
            const state = cloneDeep(prev) as any;
            state.todoList.todos.unshift(mResult.data.createTodo);
            return state;
          },
        },
      })).then(() => {
        return client.query({ query });
      }).then((newResult: any) => {
        subscriptionHandle.unsubscribe();

        // There should be one more todo item than before
        expect(newResult.data.todoList.todos.length).toBe(4);

        // Since we used `prepend` it should be at the front
        expect(newResult.data.todoList.todos[0].text).toBe(
          'This one was created with a mutation.',
        );
      }).then(resolve, reject);
    });

    itAsync('does not fail if the query did not complete correctly', (resolve, reject) => {
      const { client, obsQuery } = setupObsQuery(reject, {
        request: { query: mutation },
        result: mutationResult,
      });
      const subs = obsQuery.subscribe({
        next: () => null,
      });
      // Cancel the query right away!
      subs.unsubscribe();
      return client.mutate({
        mutation,
        updateQueries: {
          todoList: (prev, options) => {
            const mResult = options.mutationResult as any;
            expect(mResult.data.createTodo.id).toBe('99');
            expect(mResult.data.createTodo.text).toBe(
              'This one was created with a mutation.',
            );

            const state = cloneDeep(prev) as any;
            state.todoList.todos.unshift(mResult.data.createTodo);
            return state;
          },
        },
      }).then(resolve, reject);
    });

    itAsync('does not fail if the query did not finish loading', (resolve, reject) => {
      const { client, obsQuery } = setupDelayObsQuery(reject, 15, {
        request: { query: mutation },
        result: mutationResult,
      });
      obsQuery.subscribe({
        next: () => null,
      });
      return client.mutate({
        mutation,
        updateQueries: {
          todoList: (prev, options) => {
            const mResult = options.mutationResult as any;
            expect(mResult.data.createTodo.id).toBe('99');
            expect(mResult.data.createTodo.text).toBe(
              'This one was created with a mutation.',
            );

            const state = cloneDeep(prev) as any;
            state.todoList.todos.unshift(mResult.data.createTodo);
            return state;
          },
        },
      }).then(resolve, reject);
    });

    itAsync('does not make next queries fail if a mutation fails', (resolve, reject) => {
      const { client, obsQuery } = setupObsQuery(
        error => { throw error },
        {
          request: { query: mutation },
          result: { errors: [new Error('mock error')] },
        },
        {
          request: { query: queryWithTypename },
          result,
        },
      );

      obsQuery.subscribe({
        next() {
          client
            .mutate({
              mutation,
              updateQueries: {
                todoList: (prev, options) => {
                  const mResult = options.mutationResult as any;
                  const state = cloneDeep(prev) as any;
                  // It's unfortunate that this function is called at all, but we are removing
                  // the updateQueries API soon so it won't matter.
                  state.todoList.todos.unshift(
                    mResult.data && mResult.data.createTodo,
                  );
                  return state;
                },
              },
            })
            .then(
              () => reject(new Error('Mutation should have failed')),
              () =>
                client.mutate({
                  mutation,
                  updateQueries: {
                    todoList: (prev, options) => {
                      const mResult = options.mutationResult as any;
                      const state = cloneDeep(prev) as any;
                      state.todoList.todos.unshift(mResult.data.createTodo);
                      return state;
                    },
                  },
                }),
            )
            .then(
              () => reject(new Error('Mutation should have failed')),
              () => obsQuery.refetch(),
            )
            .then(resolve, reject);
        },
      });
    });

    itAsync('error handling in reducer functions', (resolve, reject) => {
      let subscriptionHandle: Subscription;
      const { client, obsQuery } = setupObsQuery(reject, {
        request: { query: mutation },
        result: mutationResult,
      });

      return obsQuery.result().then(() => {
        // we have to actually subscribe to the query to be able to update it
        return new Promise(resolve => {
          const handle = client.watchQuery({ query });
          subscriptionHandle = handle.subscribe({
            next(res) {
              resolve(res);
            },
          });
        });
      }).then(() => client.mutate({
        mutation,
        updateQueries: {
          todoList: () => {
            throw new Error(`Hello... It's me.`);
          },
        },
      })).then(() => {
        subscriptionHandle.unsubscribe();
        reject("should have thrown");
      }, error => {
        subscriptionHandle.unsubscribe();
        expect(error.message).toBe(`Hello... It's me.`);
      }).then(resolve, reject);
    });
  });

  itAsync('does not fail if one of the previous queries did not complete correctly', (resolve, reject) => {
    const variableQuery = gql`
      query Echo($message: String) {
        echo(message: $message)
      }
    `;

    const variables1 = {
      message: 'a',
    };

    const result1 = {
      data: {
        echo: 'a',
      },
    };

    const variables2 = {
      message: 'b',
    };

    const result2 = {
      data: {
        echo: 'b',
      },
    };

    const resetMutation = gql`
      mutation Reset {
        reset {
          echo
        }
      }
    `;

    const resetMutationResult = {
      data: {
        reset: {
          echo: '0',
        },
      },
    };

    const client = new ApolloClient({
      link: mockSingleLink({
        request: { query: variableQuery, variables: variables1 } as any,
        result: result1,
      }, {
        request: { query: variableQuery, variables: variables2 } as any,
        result: result2,
      }, {
        request: { query: resetMutation } as any,
        result: resetMutationResult,
      }).setOnError(reject),
      cache: new InMemoryCache({ addTypename: false }),
    });

    const watchedQuery = client.watchQuery({
      query: variableQuery,
      variables: variables1,
    });

    const firstSubs = watchedQuery.subscribe({
      next: () => null,
      error: reject,
    });

    // Cancel the query right away!
    firstSubs.unsubscribe();

    subscribeAndCount(reject, watchedQuery, (count, result) => {
      if (count === 1) {
        expect(result.data).toEqual({ echo: "a" });
      } else if (count === 2) {
        expect(result.data).toEqual({ echo: "b" });
        client.mutate({
          mutation: resetMutation,
          updateQueries: {
            Echo: () => {
              return { echo: "0" };
            },
          },
        });
      } else if (count === 3) {
        expect(result.data).toEqual({ echo: "0" });
        resolve();
      }
    });

    watchedQuery.refetch(variables2);
  });

  itAsync('allows mutations with optional arguments', (resolve, reject) => {
    let count = 0;

    const client = new ApolloClient({
      cache: new InMemoryCache({ addTypename: false }),
      link: ApolloLink.from([
        ({ variables }: any) =>
          new Observable(observer => {
            switch (count++) {
              case 0:
                expect(variables).toEqual({ a: 1, b: 2 });
                observer.next({ data: { result: 'hello' } });
                observer.complete();
                return;
              case 1:
                expect(variables).toEqual({ a: 1, c: 3 });
                observer.next({ data: { result: 'world' } });
                observer.complete();
                return;
              case 2:
                expect(variables).toEqual({
                  a: undefined,
                  b: 2,
                  c: 3,
                });
                observer.next({ data: { result: 'goodbye' } });
                observer.complete();
                return;
              case 3:
                expect(variables).toEqual({});
                observer.next({ data: { result: 'moon' } });
                observer.complete();
                return;
              default:
                observer.error(new Error('Too many network calls.'));
                return;
            }
          }),
      ] as any),
    });

    const mutation = gql`
      mutation($a: Int!, $b: Int, $c: Int) {
        result(a: $a, b: $b, c: $c)
      }
    `;

    Promise.all([
      client.mutate({
        mutation,
        variables: { a: 1, b: 2 },
      }),
      client.mutate({
        mutation,
        variables: { a: 1, c: 3 },
      }),
      client.mutate({
        mutation,
        variables: { a: undefined, b: 2, c: 3 },
      }),
      client.mutate({
        mutation,
      }),
    ])
      .then(() => {
        expect((client.cache as InMemoryCache).extract()).toEqual({
          ROOT_MUTATION: {
            __typename: 'Mutation',
            'result({"a":1,"b":2})': 'hello',
            'result({"a":1,"c":3})': 'world',
            'result({"b":2,"c":3})': 'goodbye',
            'result({})': 'moon',
          },
        });
        resolve();
      })
      .catch(reject);
  });

  itAsync('allows mutations with default values', (resolve, reject) => {
    let count = 0;

    const client = new ApolloClient({
      cache: new InMemoryCache({ addTypename: false }),
      link: ApolloLink.from([
        ({ variables }: any) =>
          new Observable(observer => {
            switch (count++) {
              case 0:
                expect(variables).toEqual({
                  a: 1,
                  b: 'water',
                });
                observer.next({ data: { result: 'hello' } });
                observer.complete();
                return;
              case 1:
                expect(variables).toEqual({
                  a: 2,
                  b: 'cheese',
                  c: 3,
                });
                observer.next({ data: { result: 'world' } });
                observer.complete();
                return;
              case 2:
                expect(variables).toEqual({
                  a: 1,
                  b: 'cheese',
                  c: 3,
                });
                observer.next({ data: { result: 'goodbye' } });
                observer.complete();
                return;
              default:
                observer.error(new Error('Too many network calls.'));
                return;
            }
          }),
      ] as any),
    });

    const mutation = gql`
      mutation($a: Int = 1, $b: String = "cheese", $c: Int) {
        result(a: $a, b: $b, c: $c)
      }
    `;

    Promise.all([
      client.mutate({
        mutation,
        variables: { a: 1, b: 'water' },
      }),
      client.mutate({
        mutation,
        variables: { a: 2, c: 3 },
      }),
      client.mutate({
        mutation,
        variables: { c: 3 },
      }),
    ])
      .then(() => {
        expect((client.cache as InMemoryCache).extract()).toEqual({
          ROOT_MUTATION: {
            __typename: 'Mutation',
            'result({"a":1,"b":"water"})': 'hello',
            'result({"a":2,"b":"cheese","c":3})': 'world',
            'result({"a":1,"b":"cheese","c":3})': 'goodbye',
          },
        });
        resolve();
      })
      .catch(reject);
  });

  itAsync('will pass null to the network interface when provided', (resolve, reject) => {
    let count = 0;

    const client = new ApolloClient({
      cache: new InMemoryCache({ addTypename: false }),
      link: ApolloLink.from([
        ({ variables }: any) =>
          new Observable(observer => {
            switch (count++) {
              case 0:
                expect(variables).toEqual({
                  a: 1,
                  b: 2,
                  c: null,
                });
                observer.next({ data: { result: 'hello' } });
                observer.complete();
                return;
              case 1:
                expect(variables).toEqual({
                  a: 1,
                  b: null,
                  c: 3,
                });
                observer.next({ data: { result: 'world' } });
                observer.complete();
                return;
              case 2:
                expect(variables).toEqual({
                  a: null,
                  b: null,
                  c: null,
                });
                observer.next({ data: { result: 'moon' } });
                observer.complete();
                return;
              default:
                observer.error(new Error('Too many network calls.'));
                return;
            }
          }),
      ] as any),
    });

    const mutation = gql`
      mutation($a: Int!, $b: Int, $c: Int) {
        result(a: $a, b: $b, c: $c)
      }
    `;

    Promise.all([
      client.mutate({
        mutation,
        variables: { a: 1, b: 2, c: null },
      }),
      client.mutate({
        mutation,
        variables: { a: 1, b: null, c: 3 },
      }),
      client.mutate({
        mutation,
        variables: { a: null, b: null, c: null },
      }),
    ])
      .then(() => {
        expect((client.cache as InMemoryCache).extract()).toEqual({
          ROOT_MUTATION: {
            __typename: 'Mutation',
            'result({"a":1,"b":2,"c":null})': 'hello',
            'result({"a":1,"b":null,"c":3})': 'world',
            'result({"a":null,"b":null,"c":null})': 'moon',
          },
        });
        resolve();
      })
      .catch(reject);
  });

  describe('store transaction updater', () => {
    const mutation = gql`
      mutation createTodo {
        # skipping arguments in the test since they don't matter
        createTodo {
          id
          text
          completed
          __typename
        }
        __typename
      }
    `;

    const mutationResult = {
      data: {
        __typename: 'Mutation',
        createTodo: {
          id: '99',
          __typename: 'Todo',
          text: 'This one was created with a mutation.',
          completed: true,
        },
      },
    };

    itAsync('analogous of ARRAY_INSERT', (resolve, reject) => {
      let subscriptionHandle: Subscription;
      const { client, obsQuery } = setupObsQuery(reject, {
        request: { query: mutation },
        result: mutationResult,
      });

      return obsQuery.result().then(() => {
        // we have to actually subscribe to the query to be able to update it
        return new Promise(resolve => {
          const handle = client.watchQuery({ query });
          subscriptionHandle = handle.subscribe({
            next(res) {
              resolve(res);
            },
          });
        });
      }).then(() => client.mutate({
        mutation,
        update: (proxy, mResult: any) => {
          expect(mResult.data.createTodo.id).toBe('99');
          expect(mResult.data.createTodo.text).toBe(
            'This one was created with a mutation.',
          );

          const id = 'TodoList5';
          const fragment = gql`
            fragment todoList on TodoList {
              todos {
                id
                text
                completed
                __typename
              }
            }
          `;

          const data: any = proxy.readFragment({ id, fragment });

          proxy.writeFragment({
            data: {
              ...data,
              todos: [mResult.data.createTodo, ...data.todos],
            },
            id,
            fragment,
          });
        },
      })).then(() => {
        return client.query({ query });
      }).then((newResult: any) => {
        subscriptionHandle.unsubscribe();

        // There should be one more todo item than before
        expect(newResult.data.todoList.todos.length).toBe(4);

        // Since we used `prepend` it should be at the front
        expect(newResult.data.todoList.todos[0].text).toBe(
          'This one was created with a mutation.',
        );
      }).then(resolve, reject);
    });

    itAsync('does not fail if optional query variables are not supplied', (resolve, reject) => {
      let subscriptionHandle: Subscription;
      const mutationWithVars = gql`
        mutation createTodo($requiredVar: String!, $optionalVar: String) {
          createTodo(requiredVar: $requiredVar, optionalVar: $optionalVar) {
            id
            text
            completed
            __typename
          }
          __typename
        }
      `;

      // the test will pass if optionalVar is uncommented
      const variables = {
        requiredVar: 'x',
        // optionalVar: 'y',
      };

      const { client, obsQuery } = setupObsQuery(reject, {
        request: {
          query: mutationWithVars,
          variables,
        },
        result: mutationResult,
      });

      return obsQuery.result().then(() => {
        // we have to actually subscribe to the query to be able to update it
        return new Promise(resolve => {
          const handle = client.watchQuery({
            query,
            variables,
          });
          subscriptionHandle = handle.subscribe({
            next(res) {
              resolve(res);
            },
          });
        });
      }).then(() => client.mutate({
        mutation: mutationWithVars,
        variables,
        update: (proxy, mResult: any) => {
          expect(mResult.data.createTodo.id).toBe('99');
          expect(mResult.data.createTodo.text).toBe(
            'This one was created with a mutation.',
          );

          const id = 'TodoList5';
          const fragment = gql`
            fragment todoList on TodoList {
              todos {
                id
                text
                completed
                __typename
              }
            }
          `;

          const data: any = proxy.readFragment({ id, fragment });

          proxy.writeFragment({
            data: {
              ...data,
              todos: [mResult.data.createTodo, ...data.todos],
            },
            id,
            fragment,
          });
        },
      })).then(() => {
        return client.query({ query });
      }).then((newResult: any) => {
        subscriptionHandle.unsubscribe();

        // There should be one more todo item than before
        expect(newResult.data.todoList.todos.length).toBe(4);

        // Since we used `prepend` it should be at the front
        expect(newResult.data.todoList.todos[0].text).toBe(
          'This one was created with a mutation.',
        );
      }).then(resolve, reject);
    });

    itAsync('does not make next queries fail if a mutation fails', (resolve, reject) => {
      const { client, obsQuery } = setupObsQuery(
        error => { throw error },
        {
          request: { query: mutation },
          result: { errors: [new Error('mock error')] },
        },
        {
          request: { query: queryWithTypename },
          result,
        },
      );

      obsQuery.subscribe({
        next() {
          client
            .mutate({
              mutation,
              update: (proxy, mResult: any) => {
                expect(mResult.data.createTodo.id).toBe('99');
                expect(mResult.data.createTodo.text).toBe(
                  'This one was created with a mutation.',
                );

                const id = 'TodoList5';
                const fragment = gql`
                  fragment todoList on TodoList {
                    todos {
                      id
                      text
                      completed
                      __typename
                    }
                  }
                `;

                const data: any = proxy.readFragment({ id, fragment });

                proxy.writeFragment({
                  data: {
                    ...data,
                    todos: [mResult.data.createTodo, ...data.todos],
                  },
                  id,
                  fragment,
                });
              },
            })
            .then(
              () => reject(new Error('Mutation should have failed')),
              () =>
                client.mutate({
                  mutation,
                  update: (proxy, mResult: any) => {
                    expect(mResult.data.createTodo.id).toBe('99');
                    expect(mResult.data.createTodo.text).toBe(
                      'This one was created with a mutation.',
                    );

                    const id = 'TodoList5';
                    const fragment = gql`
                      fragment todoList on TodoList {
                        todos {
                          id
                          text
                          completed
                          __typename
                        }
                      }
                    `;

                    const data: any = proxy.readFragment({ id, fragment });

                    proxy.writeFragment({
                      data: {
                        ...data,
                        todos: [mResult.data.createTodo, ...data.todos],
                      },
                      id,
                      fragment,
                    });
                  },
                }),
            )
            .then(
              () => reject(new Error('Mutation should have failed')),
              () => obsQuery.refetch(),
            )
            .then(resolve, reject);
        },
      });
    });

    itAsync('error handling in reducer functions', (resolve, reject) => {
      let subscriptionHandle: Subscription;
      const { client, obsQuery } = setupObsQuery(reject, {
        request: { query: mutation },
        result: mutationResult,
      });

      return obsQuery.result().then(() => {
        // we have to actually subscribe to the query to be able to update it
        return new Promise(resolve => {
          const handle = client.watchQuery({ query });
          subscriptionHandle = handle.subscribe({
            next(res) {
              resolve(res);
            },
          });
        });
      }).then(() => client.mutate({
        mutation,
        update: () => {
          throw new Error(`Hello... It's me.`);
        },
      })).then(() => {
        subscriptionHandle.unsubscribe();
        reject("should have thrown");
      }, error => {
        subscriptionHandle.unsubscribe();
        expect(error.message).toBe(`Hello... It's me.`);
      }).then(resolve, reject);
    });

    itAsync('mutate<MyType>() data should never be `undefined` in case of success', (resolve, reject) => {

      const mutation = gql`
        mutation Foo {
          foo {
            bar
          }
        }
      `;

      const result1 = {
        data: {
          foo: {
            bar: 'a',
          },
        },
      };

      const client = new ApolloClient({
        link: mockSingleLink({
          request: {query: mutation} as any,
          result: result1,

        }).setOnError(reject),
        cache: new InMemoryCache({addTypename: false}),
      });

      client.mutate<{ foo: { bar: string; }; }>({
        mutation: mutation,
      }).then(result => {
        // This next line should **not** raise "TS2533: Object is possibly 'null' or 'undefined'.", even without `!` operator
        if (result.data!.foo.bar) {
          resolve();
        }
      }, reject);
    })
  });
});
