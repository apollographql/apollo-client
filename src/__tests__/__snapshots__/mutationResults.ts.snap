// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation results should write results to cache according to errorPolicy 1`] = `Object {}`;

exports[`mutation results should write results to cache according to errorPolicy 2`] = `
Object {
  "Person:{\\"name\\":\\"Jenn Creighton\\"}": Object {
    "__typename": "Person",
    "name": "Jenn Creighton",
  },
  "ROOT_MUTATION": Object {
    "__typename": "Mutation",
    "newPerson({\\"name\\":\\"Jenn Creighton\\"})": Object {
      "__ref": "Person:{\\"name\\":\\"Jenn Creighton\\"}",
    },
  },
}
`;

exports[`mutation results should write results to cache according to errorPolicy 3`] = `
Object {
  "Person:{\\"name\\":\\"Ellen Shapiro\\"}": Object {
    "__typename": "Person",
    "name": "Ellen Shapiro",
  },
  "Person:{\\"name\\":\\"Jenn Creighton\\"}": Object {
    "__typename": "Person",
    "name": "Jenn Creighton",
  },
  "ROOT_MUTATION": Object {
    "__typename": "Mutation",
    "newPerson({\\"name\\":\\"Ellen Shapiro\\"})": Object {
      "__ref": "Person:{\\"name\\":\\"Ellen Shapiro\\"}",
    },
    "newPerson({\\"name\\":\\"Jenn Creighton\\"})": Object {
      "__ref": "Person:{\\"name\\":\\"Jenn Creighton\\"}",
    },
  },
}
`;
